### ISSUE TITLE
broadr segfault w/ thnmax

### ISSUE THREAD

**Issue by walshjon** [njoy/NJOY2016#79]
With the 2016.32 master branch, if thnmax is set "too high," broadr will segfault (stack trace below, in addition to input).  I haven't pinpointed the issue, but it appears to occur when the value of abs(thnmx) >= the second-to-greatest energy in the ENDF-6 file (e.g., 19.5 MeV for ENDF/B-VIII H-1)
Input:
```
moder
20 -21
reconr
-21 -22
'endfb80 PENDF'
125 0
0.01 0.0 0.01
0
moder
-22 32
moder
32 -22
broadr
-21 -22 -23
125 1 0 0 0.0
0.01 -19500000.0 0.01
0.0
0
moder
-23 35
acer
20 35 0 36 0
1 1 1 .00
''
125 0.0
1 0
/
acer
20 36 0 37 0
7 1 1 .00
''
stop
```
Stack trace:
```
Program received signal SIGSEGV: Segmentation fault - invalid memory reference.

Backtrace for this error:
#0  0x2aaafcb6f26f in ???
#1  0x2aaaaae60269 in bsigma
	at NJOY2016/src/broadr.f90:1550
#2  0x2aaaaae62cc4 in broadn
	at NJOY2016/src/broadr.f90:1376
#3  0x2aaaaae65965 in bfile3
	at NJOY2016/src/broadr.f90:1217
#4  0x2aaaaae6b9e9 in __broadm_MOD_broadr
	at NJOY2016/src/broadr.f90:1043
#5  0x401c86 in njoy
	at NJOY2016/src/main.f90:199
#6  0x401ef2 in main
	at NJOY2016/src/main.f90:112
Segmentation fault (core dumped)
```

**Comment by whaeck** [comment]
Interesting. I will have a look to see what is going on here. Thank you for putting this up as an issue. Doing Doppler broadening up to those energies is not something that we would advise. By default, we would stop at 1 MeV. We have never seen this problem because of that. Either way, I will have a look and figure out what is going on.

**Comment by whaeck** [comment]
So this problem revealed two different issues related to a high value of thnmax close to the upper energy value of the energy range leading to crashes. This happens when the actual temperature is zero (in that case, broadr will only thin the cross sections) and non zero (in which case broadr will apply broadening). Both cases need their own fix (in the thinb function for the first one and the bsigma function for the second one). The cause of the problem is due to the absence of range checking on the arrays containing the data we're working with.

Now, what amazes me with the above input file (in which the actual temperature is 0.0), the stack trace listed is what I get when I use a non zero temperature (e.g. 0.1 K). Either way, there is a problem for both cases.

For the thinning case (temperature equal to zero), the code gets into trouble due to a 'roundoff' at the upper end of the energy range. thnmax is set to 19.5 MeV but inside the code it is actually equal to 19.500000x. So, when comparing the before last energy point in the H1 grid (exactly 19.5 MeV) to this value, broadr thinks it is not done yet with thinning. Now, because there is no range checking in the thinb function that does the thinning, broadr continues thinning an energy array consisting of 20 MeV followed by all zeros. This can be solved by slightly changing the check to see if we are done thinning: instead of comparing the last energy value of the grid, we compare that point multiplied by 1+eps (we have to do this in two places).

For the broadening case, the error occurs in the bsigma function for the same reason as mentioned above: there is no range checking. bsigma contains this piece of code (around line 1550):
```
   k=kn
   do while (en.ge.e(k+1))
      k=k+1
   enddo
```
In this case kn is equal to 302 when it breaks down. 302 is the index of the last point of the energy array for H1. So the element e(kn)=20 MeV and e(k+1) equals 0.0. As a result, the while statement is always true (as e(k+1) is always 0) until it goes over the size of the array into unallocated memory - which causes the crash seen above. This can be solved by simply checking if the next energy point is actually equal to or larger than the curent one before going into the do loop:
```
   k=kn
   if (e(k+1).ge.e(k)) then
      do while (en.ge.e(k+1))
         k=k+1
      enddo
   endif
```

Both fixes now run njoy to completion when the temperature in the above input is 0.0 or 0.1.

These fixes are available in a branch: feature/broadr-endf8. I will issue a pull request to put these fixes into the master, but not until I have added two tests (tem2 equal to 0.0 and 0.1) and maybe some warning messages concerning thnmax to inform the user that thnmax may be to high.

**Comment by whaeck** [comment]
These fixes are now in NJOY 2016.37
